crates/task-center/src/lib.rs

```diff
@@ -0,0 +1,554 @@
+    join_handle: Mutex<Option<JoinHandle<()>>>,
+}
+
+task_local! {
+    // This is a cancellation token which will be cancelled when a task needs to shut down.
+    static CANCEL_TOKEN: CancellationToken;
+
+    // Tasks have self-reference.
+    static CURRENT_TASK: Arc<Task>;
+
+    // Current task center
+    static CURRENT_TASK_CENTER: TaskCenter;
+}
+
+/// This wrapper function runs in a newly-spawned task. It initializes the
+/// task-local variables and calls the payload function.
+async fn wrapper<F>(
+    task_center: TaskCenter,
+    task_id: TaskId,
+    kind: TaskKind,
+    task: Arc<Task>,
+    cancel_token: CancellationToken,
+    shutdown_process_on_error: bool,
+    future: F,
+) where
+    F: Future<Output = anyhow::Result<()>> + Send + 'static,
+{
+    debug!(kind = ?kind, name = ?task.name, "Starting task {}", task_id);
+
+    let result = CURRENT_TASK_CENTER
+        .scope(
+            task_center.clone(),
+            CANCEL_TOKEN.scope(
+                cancel_token,
+                CURRENT_TASK.scope(task, {
+                    // We use AssertUnwindSafe here so that the wrapped function
+                    // doesn't need to be UnwindSafe. We should not do anything after
+                    // unwinding that'd risk us being in unwind-unsafe behavior.
+                    AssertUnwindSafe(future).catch_unwind()
+                }),
+            ),
+        )
+        .await;
+    task_center
+        .on_finish(result, kind, task_id, shutdown_process_on_error)
+        .await;
+}
+
+/// The current task-center task kind. This returns None if we are not in the scope
+/// of a task-center task.
+pub fn current_task_kind() -> Option<TaskKind> {
+    CURRENT_TASK.try_with(|ct| ct.kind).ok()
+}
+
+/// The current task-center task Id. This returns None if we are not in the scope
+/// of a task-center task.
+pub fn current_task_id() -> Option<TaskId> {
+    CURRENT_TASK.try_with(|ct| ct.id).ok()
+}
+
+/// The current partition Id associated to the running task-center task.
+pub fn current_task_partition_id() -> Option<PartitionId> {
+    CURRENT_TASK.try_with(|ct| ct.partition_id).ok().flatten()
+}
+
+/// Get the current task center. Use this to spawn tasks on the current task center.
+/// This must be called from within a task-center task.
+pub fn task_center() -> TaskCenter {
+    CURRENT_TASK_CENTER
+        .try_with(|t| t.clone())
+        .expect("task_center() called in a task-center task")
+}
+
+/// A Future that can be used to check if the current task has been requested to
+/// shut down.
+pub async fn cancellation_watcher() {
+    let token = CANCEL_TOKEN
+        .try_with(|t| t.clone())
+        .expect("cancellation_watcher() called in a task-center task");
+
+    token.cancelled().await;
+}
+
+/// Clone the current task's cancellation token, which can be moved across tasks.
+///
+/// When the task which is currently executing is shutdown, the cancellation token will be
+/// cancelled.
```

Comment:
Do I understand this invariant correctly that the cancellation token will be canceled if the corresponding task is shut down? Also if the task shuts down normally?